### PR TITLE
Adviser  

### PR DESCRIPTION
**Pull request by JacobLeach**
Needs more testing and perhaps some corrections for three card ace hands

### CODE REVIEW

**Comment by JacobLeach** [comment]
Google led me to fuzz testing. i'm going to try and make it so i can automatically throw it all sorts of inputs and check for an exception

**Comment by JacobLeach** [comment]
Added a Fuzz Unit Test. Running it 10M times now.

**Comment by JacobLeach** [comment]
Passed 10M, running 1B.

**Comment by JacobLeach** [comment]
I'm assuming I shouldn't have to check for the hand being bust or blackjack right? I should assume that the hands passed are correct?

**Comment by JacobLeach** [comment]
Ok the 1B didn't finished (after hours lol) but I've got to assume it hit all the possible hands.

File: src/charlie/card/Card.java [hunk 1]
```diff
@@ -108,7 +109,33 @@ else if(rank == KING)
         else
             return "?";
     }
-    
+   
+    @Override
+    public boolean equals(Object obj)
```
Comment (travisby) [inline src/charlie/card/Card.java:139]: Are we sure that the previous object equals() wasn't used in any of Charlie's other code?

File: src/charlie/plugins/Advisor.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
```
Comment (travisby) [inline src/charlie/plugins/Advisor.java]: License header :P

File: src/charlie/plugins/Advisor.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                        {null, S, S, S, S, S, S, S, S, S, S}, //8
+                        {null, S, S, S, S, S, S, S, S, S, S}, //9
+                        {null, S, S, S, S, S, S, S, S, S, S}, //10
+                        };
+    @Override
+    public Play advise(Hand myHand, Card upCard)
+    {
+        //Set default return to SPLIT for testing
+        //If it says SPLIT, it did not have any advice
+        Play toReturn = Play.SPLIT;
```
Comment (travisby) [inline src/charlie/plugins/Advisor.java:75]: Why is it using SPLIT for this?

File: test/charlie/test/AdvisorTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    private Advisor advisor;
+
+    @Before
+    public void setup()
+    {
+        advisor = new Advisor();
+    }
+
+    @Test
+    public void fuzzTest()
```
Comment (travisby) [inline test/charlie/test/AdvisorTest.java:26]: niiiice

File: test/charlie/test/AdvisorTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        Hand hand = null;
+        Card upCard = null;
+
+        try
+        {
+            for (int i = 0; i < RUN_NUMBER; i++)
+            {
+                hand = new Hand(new Hid(Seat.YOU));
+                hand.hit(RandomCardFactory.getRandomCard());
+                hand.hit(RandomCardFactory.getRandomCard());
```
Comment (travisby) [inline test/charlie/test/AdvisorTest.java]: RandomCardFactory is epic

**Comment by travisby** [comment]
Cursory reading looks really good.  I'll run the tests and try to come up with edge cases tomorrow night, and then accept the PR. Added a few comments to some of the lines if you could take a look at them.

File: src/charlie/card/Card.java (hunk 1 above)
Comment (JacobLeach) [inline src/charlie/card/Card.java:139]: Good point. I'll check. The hash function that netbeans made doesn't seem to work anyways.

File: src/charlie/plugins/Advisor.java (hunk 2 above)
Comment (JacobLeach) [inline src/charlie/plugins/Advisor.java]: Fuck netbeans and its autofill. i'll get rid of that

File: src/charlie/plugins/Advisor.java (hunk 3 above)
Comment (JacobLeach) [inline src/charlie/plugins/Advisor.java:75]: If it ever returns SPLIT, its wrong. I should test for that!

File: src/charlie/card/Card.java (hunk 1 above)
Comment (JacobLeach) [inline src/charlie/card/Card.java:139]: It does not seem to be used anywhere.

Comment (travisby) [inline src/charlie/card/Card.java:139]: :+1:

File: src/charlie/plugins/Advisor.java (hunk 3 above)
Comment (travisby) [inline src/charlie/plugins/Advisor.java:75]: I mean it doesn't seem semantically correct.  We should probably remove SPLIT, and put it Play.ERR or something similar (and then throw)

Comment (JacobLeach) [inline src/charlie/plugins/Advisor.java:75]: Yea I know. Its just that I didn't want to add another option to the enum because that would make the plugin only work on the modified Charlie. If only java had macros because then i'd just make it different for debug mode!

File: src/charlie/card/Card.java (hunk 1 above)
Comment (JacobLeach) [inline src/charlie/card/Card.java:139]: +1 to grep

File: src/charlie/plugins/Advisor.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                        {null, H, H, H, D, D, D, H, H, H, H}, //4
+                        {null, H, H, H, D, D, D, H, H, H, H}, //5
+                        {null, H, H, D, D, D, D, H, H, H, H}, //6
+                        {null, H, S, D, D, D, D, S, S, H, H}, //7
+                        {null, S, S, S, S, S, S, S, S, S, S}, //8
+                        {null, S, S, S, S, S, S, S, S, S, S}, //9
+                        {null, S, S, S, S, S, S, S, S, S, S}, //10
+                        };
+    @Override
+    public Play advise(Hand myHand, Card upCard)
```
Comment (JacobLeach) [inline src/charlie/plugins/Advisor.java:71]: Should I be checking the parameters or just assuming that I am passed a correct hand?? I assume i could assume that the hand is correct at this level or abstraction.

**Comment by travisby** [comment]
does any more of this code depend on SPLIT never being used?  We must go back and reimplement split.

**Comment by JacobLeach** [comment]
Yea. I have to redo it since there are now hand combinations that I do not account for. Before I just treated hands with aces as special. Now I have to check if it is a special hand, and it is not then do it by value. There is a NONE so we can use that instead.

**Comment by travisby** [comment]
schweet

**Comment by travisby** [comment]
testing

**Comment by travisby** [comment]
Good with the latest commits!

:shipit:
